presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java

```diff
@@ -1114,35 +1117,51 @@ public PlanWithProperties visitUnion(UnionNode node, Context context)
                 }
             }
 
-            PlanNode result = null;
-            if (!partitionedChildren.isEmpty()) {
-                // add an exchange above partitioned inputs and fold it into the
-                // set of unpartitioned inputs
-                // NOTE: this must provide the explicit imput mapping as unions may drop or duplicate symbols
+            PlanNode result;
+            if (!partitionedChildren.isEmpty() && unpartitionedChildren.isEmpty()) {
+                // add an gathering exchange above partitioned inputs
                 result = new ExchangeNode(
                         idAllocator.getNextId(),
                         GATHER,
                         REMOTE,
                         new PartitioningScheme(Partitioning.create(SINGLE_DISTRIBUTION, ImmutableList.of()), node.getOutputSymbols()),
                         partitionedChildren,
                         partitionedOutputLayouts);
-
-                unpartitionedChildren.add(result);
-                unpartitionedOutputLayouts.add(result.getOutputSymbols());
             }
+            else if (!unpartitionedChildren.isEmpty()) {
+                if (!partitionedChildren.isEmpty()) {
+                    // add an gathering exchange above partitioned inputs and fold it into the set of unpartitioned inputs
+                    // NOTE: new symbols for ExchangeNode output are required in order to keep plan logically correct with new local union below
+
+                    List<Symbol> exchangeOutputLayout = node.getOutputSymbols().stream()
```

Comment:
As I understand in case there are just `partitionedChildren` just the `ExchangeNode` needs to be added. This node has to use the same output symbols as the initial `UnionNode`, because those are referenced by downstream node.

When there are some `partitionedChildren` and `unpartitionedChildren`, the exchange node is an input to local `UnionNode`. In order for the local `UnionNode` to be symbolically correct, the `ExchangeNode` has to use new set of output symbols.